WIP Testing with MSW & React testing library

JMeteor opened the pull request.

JMeteor committed:
WIP Test InvoiceForm

DBulinski commented on src/components/InvoiceForm.test.ts:
```diff
@@ -0,0 +1,10 @@
+
+describe('InvoiceForm', () => {
+  it('renders without crashing', () => {
+    render(<InvoiceForm mode="view" onSubmit={vi.fn()} resolver={vi.fn()} />);
```
I'd rather test it one level higher (EditInvoice and AddInvoice pages). But I understand this is not ready yet :)

DBulinski commented on src/services/invoices/invoicesService.test.ts:
```diff
@@ -0,0 +1,96 @@
+const MOCK_INVOICE = mockInvoice;
+
+describe('invoicesService', () => {
+  beforeAll(() => server.listen());
```
You can setup this in setupFiles https://vitest.dev/config/#setupfiles

DBulinski commented on src/services/invoices/invoicesService.test.ts:
```diff
@@ -0,0 +1,96 @@
+
+  describe('fetchAllInvoices', () => {
+    test('returns invoices data when invoice exists', async () => {
+      const data = await InvoicesService.fetchAllInvoices();
```
These tests mostly test if msw is working correctly :D I'd rather treat it as an implementation detail, user doesn't care if you are calling GET / POST or PUT, he/she just wants to get and modify data :D. So I'd rather suggest testing the whole InvoiceList / AddInvoice / EditInvoice page which calls MSW and gets data instead.

DBulinski commented on src/services/invoices/invoicesService.test.ts:
```diff
@@ -0,0 +1,96 @@
+        name: 'Invoice 2',
+      };
+
+      //TODO: How to mock updated invoice data?
```
Not sure I understand the question :D

DBulinski commented on src/services/invoices/invoicesService.test.ts:
```diff
@@ -0,0 +1,96 @@
+
+  describe('deleteInvoice', () => {
+    test('deletes the invoice', async () => {
+      //TODO: How to test deleting an invoice?
```
If you want to, I'd try to get invoice with id '1' after the deleteInvoice call, and expect it to return 404. But if you test it in InvoiceList, you can just verify if it disappears from table, and then you don't need to care about the actual API call, which is an implementation detail. If you change the impl to use graphql instead of rest, and change the service to handle it, you'd expect tests to pass, as the app behaves in the same fashion.

DBulinski commented on src/utils/test-utils.tsx:
```diff
@@ -0,0 +1,21 @@
+import React, { ReactElement } from 'react';
```
I'd suggest:
{ ReactElement, ReactNode } or * as React 
But it's minor

DBulinski commented on src/utils/test-utils.tsx:
```diff
@@ -0,0 +1,21 @@
+
+const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
+  return (
+    <LocalizationProvider dateAdapter={AdapterDateFns}>
```
I think you are missing QueryProvider

JMeteor committed:
Review fixes

DBulinski commented on src/test/setup.ts:
```diff
@@ -0,0 +1,6 @@
+import { afterAll, afterEach, beforeAll } from 'vitest';
+import { server } from '../mocks/server.ts';
+
+beforeAll(() => server.listen());
```
```bash
  "dev:mocks": cross-env ENABLE_MOCKS=true npx dev
```


worker.ts
```ts
export const worker = setupWorker(...handlers)
```

main.tsx
```ts
 import {worker} from './mocks/worker.ts'

const renderApp = () => {}

 if(import.env.ENABLE_MOCKS === 'true') {
   worker.start().then(renderApp)
 } else {
renderApp()
}
```

DBulinski committed:
review

JMeteor committed:
add-invoice page tests

JMeteor committed:
Upgrade vite

JMeteor committed:
Add invoice tests and navigate to view after save

JMeteor committed:
Add env variable for enabling msw

JMeteor committed:
Date and amount listing display fixes

JMeteor committed:
Refactor test-utils for readability and efficiency

The changes in this commit streamline the code in the test-utils file, reducing redundancy and promoting efficiency. Specifically, the 'customRender' and 'renderWithRouter' functions have been adjusted to enhance the clarity of their parameters, allowing for greater readability and easier maintenance going forward.

JMeteor committed:
Refactor tests for adding, viewing, and listing invoices

DBulinski commented on src/routes/add-invoice.test.tsx:
```diff
@@ -0,0 +1,121 @@
+    ) as HTMLInputElement;
+    await user.type(orderItemPriceField, '100');
+
+    expect(nameField.value).toBe('Test Company');
```
I would say there is no need to test such behaviour. This is handled by the browser, so I'm pretty sure it is really well tested :D

DBulinski commented on src/routes/add-invoice.test.tsx:
```diff
@@ -0,0 +1,121 @@
+    const saveButton = screen.getByText('Save');
+    await user.click(saveButton);
+
+    expect(screen.findByText('Test Company')).toBeInTheDocument();
```
You can check if the user was redirected to a right page here.

DBulinski commented on src/routes/add-invoice.test.tsx:
```diff
@@ -0,0 +1,121 @@
+import App from '../App.tsx';
+
+describe('AddInvoicePage', () => {
+  const { user } = renderWithRouter(<App />);
```
If you are going to render whole App, then there is no need to use `renderWithRouter`, you can use simple `render`. RenderWithRouter wraps it with additional Providers, which duplicates all contexts (Router, QueryClient etc)

DBulinski commented on src/routes/add-invoice.test.tsx:
```diff
@@ -0,0 +1,121 @@
+    expect(orderItemAmountField.value).toBe('1');
+    expect(orderItemUnitField.value).toBe('Test');
+    expect(orderItemTaxField.value).toBe('23');
+    expect(orderItemPriceField.value).toBe('100');
```
If you want to verify, if the data was sent correctly, you can spy on a method and check the data shape. But I'd rather reuse zod in the MSW handler. That way if the data is corrupted, msw will return an error and the test will fail
